src/lib.rs

```diff
@@ -1467,3 +1433,96 @@ impl Disposition {
         }
     }
 }
+
+// Request helper type for NATS request
+struct Request {
+    reply_to: Subject,
+    wrapped_client: Arc<Mutex<SyncClient>>,
+}
+
+impl Request {
+    async fn new(wrapped_client: Arc<Mutex<SyncClient>>) -> Result<Self> {
+        let inbox_uuid = Uuid::new_v4();
+
+        let client = wrapped_client.lock().await;
+        let request_inbox = inbox_uuid.to_simple();
+        let reply_to: Subject = format!(
+            "{}.{}.{}",
+            util::INBOX_PREFIX,
+            client.request_base_inbox,
+            request_inbox
+        )
+        .parse()?;
+
+        Ok(Self {
+            reply_to,
+            wrapped_client: wrapped_client.clone(),
+        })
+    }
+
+    async fn call(
+        &mut self,
+        subject: &Subject,
+        payload: &[u8],
+        duration: Option<Duration>,
+    ) -> Result<Msg> {
+        let mut rx = {
+            let mut client = self.wrapped_client.lock().await;
+
+            // Only subscribe to the wildcard subscription when requested once!
+            if client.request_wildcard_subscription.is_none() {
+                let global_reply_to =
+                    format!("{}.{}.*", util::INBOX_PREFIX, client.request_base_inbox).parse()?;
+                let (sid, rx) = client.subscribe(&global_reply_to, 1024).await?;
+                client.request_wildcard_subscription = Some(sid);
+
+                // Spawn the task that watches the request wildcard receiver.
+                tokio::spawn(SyncClient::request_wildcard_handler(
+                    self.wrapped_client.clone(),
+                    rx,
+                ));
+            }
+
+            let (tx, rx) = mpsc::channel(1);
+            client
+                .request_inbox_mapping
+                .insert(self.reply_to.clone(), tx);
+            client
+                .publish_with_reply(subject, &self.reply_to, payload)
+                .await?;
+
+            rx
+        };
+
+        // Use a timeout duration if it was provided by the caller.
+        let next_message = match duration {
+            Some(duration) => tokio::time::timeout(duration, rx.next()).await?,
+            None => rx.next().await,
+        };
+
+        // Make sure we clean up on error (don't leave a dangling request
+        // inbox mapping reference. Adding an extra mutex here seems fine
+        // since this is the error path.
+        match next_message {
+            Some(response) => Ok(response),
+            None => Err(Error::NoResponse),
```

Comment:
> I also think it might be worth setting a flag on error so we only attempt to cleanup if we absolutely need to.

Would you mind adding this here? I dont know how much difference it will make, but think it is a good idea to be safe.

Reply:
Done. It would be nice to make the request callable over and over. I can reset the `request_inbox_mapping_was_removed` flag but it seems a bit wrong. Any ideas here?

Reply:
:thinking: that is a good point. If `call` is called multiple times I am not sure what the behavior should be. What happens if we receive a message from a previous request? For now we could just make `call` take ownership of `self` which only allows it to be called once and we can also remove the `request_inbox_mapping_was_removed` and the `Drop` impl and just do the inbox cleanup in `call`. If we ever made `Request` public we would want to allow `call`to be called multiple times and add the `Drop` impl back in.

Reply:
I like the idea of `call` taking ownership. At the very least it lets the compiler enforce hidden behavior. I pushed the change. If we always set a timeout then we can move the drop logic into `call` and remove the drop impl, but given it's possible to have no timeout, we need the drop to handle a request that was tossed away for cleanup.